File: backend/app/seed.py
"""Seed the database with default categories and sample data."""
from app.database import SessionLocal
from app.models.user import User
from app.models.account import Account
from app.models.category import Category
from app.models.transaction import Transaction
from app.models.budget import Budget
from app.services.auth import hash_password
from datetime import date, timedelta
import random

DEFAULT_CATEGORIES = [
    # (name, icon, color, is_income, children)
    ("Income", "", "#10b981", True, [
        ("Salary", "", "#10b981", True),
        ("Freelance", "", "#34d399", True),
        ("Investments", "", "#6ee7b7", True),
        ("Other Income", "", "#a7f3d0", True),
    ]),
    ("Housing", "", "#6366f1", False, [
        ("Rent/Mortgage", "", "#6366f1", False),
        ("Utilities", "", "#818cf8", False),
        ("Insurance", "", "#a5b4fc", False),
        ("Maintenance", "", "#c7d2fe", False),
    ]),
    ("Food", "", "#f59e0b", False, [
        ("Groceries", "", "#f59e0b", False),
        ("Restaurants", "", "#fbbf24", False),
        ("Coffee", "", "#fcd34d", False),
    ]),
    ("Transportation", "", "#ef4444", False, [
        ("Gas", "", "#ef4444", False),
        ("Public Transit", "", "#f87171", False),
        ("Car Payment", "", "#fca5a5", False),
        ("Car Insurance", "", "#fecaca", False),
    ]),
    ("Entertainment", "", "#8b5cf6", False, [
        ("Streaming", "", "#8b5cf6", False),
        ("Games", "", "#a78bfa", False),
        ("Events", "", "#c4b5fd", False),
    ]),
    ("Health", "", "#ec4899", False, [
        ("Medical", "", "#ec4899", False),
        ("Dental", "", "#f472b6", False),
        ("Gym", "", "#f9a8d4", False),
    ]),
    ("Shopping", "", "#14b8a6", False, [
        ("Clothing", "", "#14b8a6", False),
        ("Electronics", "", "#2dd4bf", False),
        ("Home Goods", "", "#5eead4", False),
    ]),
    ("Personal", "", "#f97316", False, [
        ("Subscriptions", "", "#f97316", False),
        ("Education", "", "#fb923c", False),
        ("Gifts", "", "#fdba74", False),
    ]),
]

SAMPLE_PAYEES = {
    "Groceries": ["Whole Foods", "Trader Joe's", "Costco", "Safeway", "Kroger"],
    "Restaurants": ["Chipotle", "Olive Garden", "Thai Express", "Pizza Hut", "Sushi Palace"],
    "Coffee": ["Starbucks", "Blue Bottle", "Peet's Coffee", "Local Cafe"],
    "Gas": ["Shell", "Chevron", "BP", "Exxon"],
    "Streaming": ["Netflix", "Spotify", "Disney+", "HBO Max"],
    "Rent/Mortgage": ["Property Management Co."],
    "Utilities": ["Electric Co.", "Water Dept.", "Internet Provider", "Gas Company"],
    "Salary": ["Employer Inc."],
}


def seed_database():
    db = SessionLocal()
    try:
        if db.query(User).count() > 0:
            print("Database already has data, skipping seed.")
            return

        # Create demo user
        user = User(
            username="demo",
            email="demo@budgetbolt.local",
            hashed_password=hash_password("demo123"),
        )
        db.add(user)
        db.flush()

        # Create accounts
        checking = Account(user_id=user.id, name="Main Checking", type="checking", currency="USD")
        savings = Account(user_id=user.id, name="Savings", type="savings", currency="USD")
        credit = Account(user_id=user.id, name="Credit Card", type="credit_card", currency="USD")
        db.add_all([checking, savings, credit])
        db.flush()

        # Create categories
        cat_map = {}
        for parent_name, parent_icon, parent_color, parent_income, children in DEFAULT_CATEGORIES:
            parent = Category(
                user_id=user.id, name=parent_name, icon=parent_icon,
                color=parent_color, is_income=parent_income,
            )
            db.add(parent)
            db.flush()
            cat_map[parent_name] = parent

            for child_name, child_icon, child_color, child_income in children:
                child = Category(
                    user_id=user.id, name=child_name, parent_id=parent.id,
                    icon=child_icon, color=child_color, is_income=child_income,
                )
                db.add(child)
                db.flush()
                cat_map[child_name] = child

        # Generate 3 months of sample transactions
        today = date.today()
        for month_offset in range(3):
            month_start = today.replace(day=1) - timedelta(days=30 * month_offset)
            month_start = month_start.replace(day=1)

            # Monthly salary
            salary_cat = cat_map.get("Salary")
            if salary_cat:
                db.add(Transaction(
                    user_id=user.id, account_id=checking.id, category_id=salary_cat.id,
                    date=month_start.replace(day=1), payee="Employer Inc.",
                    amount=5000.00, notes="Monthly salary",
                ))

            # Random expenses
            expense_categories = [
                ("Groceries", -40, -150, 4),
                ("Restaurants", -15, -60, 3),
                ("Coffee", -4, -8, 5),
                ("Gas", -30, -60, 2),
                ("Streaming", -12, -16, 1),
                ("Utilities", -80, -200, 1),
            ]

            for cat_name, min_amt, max_amt, count in expense_categories:
                cat = cat_map.get(cat_name)
                if not cat:
                    continue
                payees = SAMPLE_PAYEES.get(cat_name, ["Misc"])
                for _ in range(count):
                    day = random.randint(1, 28)
                    txn_date = month_start.replace(day=day)
                    db.add(Transaction(
                        user_id=user.id,
                        account_id=random.choice([checking.id, credit.id]),
                        category_id=cat.id,
                        date=txn_date,
                        payee=random.choice(payees),
                        amount=round(random.uniform(min_amt, max_amt), 2),
                        notes="",
                    ))

            # Create budgets for this month
            month_str = month_start.strftime("%Y-%m")
            budget_items = [
                ("Groceries", 600), ("Restaurants", 200), ("Coffee", 50),
                ("Gas", 150), ("Streaming", 30), ("Utilities", 300),
                ("Rent/Mortgage", 1500), ("Clothing", 100), ("Entertainment", 100),
            ]
            for cat_name, amount in budget_items:
                cat = cat_map.get(cat_name)
                if cat:
                    db.add(Budget(
                        user_id=user.id, category_id=cat.id,
                        month=month_str, assigned=amount,
                    ))

        db.commit()
        print("Database seeded successfully!")
        print("Demo login: username=demo, password=demo123")

    finally:
        db.close()


if __name__ == "__main__":
    from app.database import Base, engine
    Base.metadata.create_all(bind=engine)
    seed_database()

File: v2/convex/seed.ts
import { internalMutation } from "./_generated/server";
import type { Id } from "./_generated/dataModel";

const DEFAULT_CATEGORIES = [
  { name: "Salary", icon: "briefcase", color: "#22c55e", isIncome: true },
  { name: "Freelance", icon: "laptop", color: "#10b981", isIncome: true },
  { name: "Investments", icon: "trending-up", color: "#06b6d4", isIncome: true },
  { name: "Housing", icon: "home", color: "#6366f1", isIncome: false },
  { name: "Transportation", icon: "car", color: "#8b5cf6", isIncome: false },
  { name: "Food & Dining", icon: "utensils", color: "#f59e0b", isIncome: false },
  { name: "Utilities", icon: "zap", color: "#ef4444", isIncome: false },
  { name: "Healthcare", icon: "heart", color: "#ec4899", isIncome: false },
  { name: "Entertainment", icon: "film", color: "#f97316", isIncome: false },
  { name: "Shopping", icon: "shopping-bag", color: "#14b8a6", isIncome: false },
  { name: "Insurance", icon: "shield", color: "#64748b", isIncome: false },
  { name: "Personal Care", icon: "smile", color: "#d946ef", isIncome: false },
  { name: "Education", icon: "book", color: "#0ea5e9", isIncome: false },
  { name: "Subscriptions", icon: "repeat", color: "#a855f7", isIncome: false },
  { name: "Gifts & Donations", icon: "gift", color: "#e11d48", isIncome: false },
];

export const seedData = internalMutation({
  handler: async (ctx) => {
    // Check if data already exists
    const existing = await ctx.db.query("categories").first();
    if (existing) {
      console.log("Seed data already exists, skipping");
      return;
    }

    // We need a user to attach data to — this is just for demo/testing
    // In production, seed data is created per-user on first login
    const users = await ctx.db.query("users").collect();
    if (users.length === 0) {
      console.log("No users found — register first, then run seed");
      return;
    }

    const userId = users[0]._id;

    // Create categories
    const categoryIds: Record<string, Id<"categories">> = {};
    for (const cat of DEFAULT_CATEGORIES) {
      const id = await ctx.db.insert("categories", {
        userId,
        name: cat.name,
        icon: cat.icon,
        color: cat.color,
        isIncome: cat.isIncome,
      });
      categoryIds[cat.name] = id;
    }

    // Create accounts
    const checkingId = await ctx.db.insert("accounts", {
      userId,
      name: "Main Checking",
      type: "checking",
      balance: 5200,
      currency: "USD",
      isActive: true,
    });

    const savingsId = await ctx.db.insert("accounts", {
      userId,
      name: "Emergency Fund",
      type: "savings",
      balance: 15000,
      currency: "USD",
      isActive: true,
    });

    const creditId = await ctx.db.insert("accounts", {
      userId,
      name: "Visa Card",
      type: "credit_card",
      balance: -1250,
      currency: "USD",
      isActive: true,
    });

    // Create sample transactions for current and previous month
    const now = new Date();
    const thisMonth = `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, "0")}`;
    const lastMonth = now.getMonth() === 0
      ? `${now.getFullYear() - 1}-12`
      : `${now.getFullYear()}-${String(now.getMonth()).padStart(2, "0")}`;

    const sampleTransactions = [
      { date: `${thisMonth}-01`, payee: "Employer Inc", amount: 4500, accountId: checkingId, categoryId: categoryIds["Salary"], notes: "Monthly salary" },
      { date: `${thisMonth}-03`, payee: "Rent Payment", amount: -1800, accountId: checkingId, categoryId: categoryIds["Housing"], notes: "Monthly rent" },
      { date: `${thisMonth}-05`, payee: "Grocery Store", amount: -125.50, accountId: creditId, categoryId: categoryIds["Food & Dining"], notes: "" },
      { date: `${thisMonth}-07`, payee: "Electric Company", amount: -95.20, accountId: checkingId, categoryId: categoryIds["Utilities"], notes: "Monthly electric" },
      { date: `${thisMonth}-08`, payee: "Netflix", amount: -15.99, accountId: creditId, categoryId: categoryIds["Subscriptions"], notes: "" },
      { date: `${thisMonth}-10`, payee: "Gas Station", amount: -45.00, accountId: creditId, categoryId: categoryIds["Transportation"], notes: "" },
      { date: `${thisMonth}-12`, payee: "Restaurant", amount: -62.30, accountId: creditId, categoryId: categoryIds["Food & Dining"], notes: "Dinner out" },
      { date: `${thisMonth}-15`, payee: "Freelance Client", amount: 800, accountId: checkingId, categoryId: categoryIds["Freelance"], notes: "Web project" },
      { date: `${lastMonth}-01`, payee: "Employer Inc", amount: 4500, accountId: checkingId, categoryId: categoryIds["Salary"], notes: "Monthly salary" },
      { date: `${lastMonth}-03`, payee: "Rent Payment", amount: -1800, accountId: checkingId, categoryId: categoryIds["Housing"], notes: "Monthly rent" },
      { date: `${lastMonth}-05`, payee: "Grocery Store", amount: -145.80, accountId: creditId, categoryId: categoryIds["Food & Dining"], notes: "" },
      { date: `${lastMonth}-10`, payee: "Internet Provider", amount: -79.99, accountId: checkingId, categoryId: categoryIds["Utilities"], notes: "" },
      { date: `${lastMonth}-15`, payee: "Amazon", amount: -89.99, accountId: creditId, categoryId: categoryIds["Shopping"], notes: "Electronics" },
      { date: `${lastMonth}-20`, payee: "Doctor Visit", amount: -150.00, accountId: checkingId, categoryId: categoryIds["Healthcare"], notes: "Annual checkup" },
    ];

    for (const tx of sampleTransactions) {
      await ctx.db.insert("transactions", {
        userId,
        accountId: tx.accountId,
        categoryId: tx.categoryId,
        date: tx.date,
        payee: tx.payee,
        amount: tx.amount,
        notes: tx.notes,
        isTransfer: false,
      });
    }

    // Create budgets for current month
    const budgetItems = [
      { category: "Housing", amount: 1800 },
      { category: "Food & Dining", amount: 400 },
      { category: "Transportation", amount: 200 },
      { category: "Utilities", amount: 200 },
      { category: "Entertainment", amount: 150 },
      { category: "Shopping", amount: 100 },
      { category: "Subscriptions", amount: 50 },
      { category: "Healthcare", amount: 100 },
    ];

    for (const item of budgetItems) {
      if (categoryIds[item.category]) {
        await ctx.db.insert("budgets", {
          userId,
          categoryId: categoryIds[item.category],
          month: thisMonth,
          assigned: item.amount,
        });
      }
    }

    console.log("Seed data created successfully!");
  },
});
